src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Base2ExponentialHistogramAggregator.cs

```diff
@@ -0,0 +1,353 @@
+
+        /// <summary>
+        /// The number of zero values in the histogram.
+        /// </summary>
+        internal long ZeroCount { get; private set; }
+
+        public override IAggregationStatistics Collect()
+        {
+            lock (this)
+            {
+                var newStats = new Base2ExponentialHistogramStatistics(
+                    Scale,
+                    ZeroCount,
+                    _sum,
+                    _count,
+                    _minMeasurement,
+                    _maxMeasurement,
+                    PositiveBuckets.ToArray());
+
+                if (_reportDeltas)
+                {
+                    // Reset the state of the aggregator.
+                    _sum = 0;
+                    _minMeasurement = double.MaxValue;
+                    _maxMeasurement = double.MinValue;
+                    ZeroCount = 0;
+                    _count = 0;
+                    PositiveBuckets.Clear();
+                }
+
+                return newStats;
+            }
+        }
+
+        public override void Update(double measurement)
+        {
+            if (!IsFinite(measurement))
+            {
+                return;
+            }
+
+            var c = measurement.CompareTo(0);
+
+            if (c < 0)
+            {
+                return; // Negative values are not supported according to the OpenTelemetry spec.
+            }
+
+            lock (this)
+            {
+                _maxMeasurement = Math.Max(_maxMeasurement, measurement);
+                _minMeasurement = Math.Min(_minMeasurement, measurement);
+                _count++;
+
+                if (c == 0)
+                {
+                    ZeroCount++;
+                    return;
+                }
+
+                _sum += measurement;
+
+                var index = MapToIndex(measurement);
+                var n = PositiveBuckets.TryIncrement(index);
+
+                if (n == 0)
+                {
+                    return;
+                }
+
+                PositiveBuckets.ScaleDown(n);
+                Scale -= n;
+
+                if (Scale < MinScale)
+                {
+                    Scale = MinScale;
+                }
+                else if (Scale > MaxScale)
+                {
+                    Scale = MaxScale;
+                }
+
+                n = PositiveBuckets.TryIncrement(index >> n);
+                Debug.Assert(n == 0, "Increment should always succeed after scale down.");
+            }
+        }
+
+        /// <summary>
+        /// Maps a finite positive IEEE 754 double-precision floating-point
+        /// number to <c>Bucket[index] = ( base ^ index, base ^ (index + 1) ]</c>,
+        /// where <c>index</c> is an integer.
+        /// </summary>
+        /// <param name="value">The value to be bucketized. Must be a finite positive number.</param>
+        /// <returns>Returns the index of the bucket.</returns>
+        public int MapToIndex(double value)
```

Comment:
If we need to copy code from OpenTelemetry, we need to follow Apache License 2.0 and include the proper copyright header.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/805dd6b4abfa18ef2706d04c30d0ed28dbc2955e/LICENSE.TXT#L100-L102

> You must retain, in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work

Reply:
Thanks!

We always include licensing information in the third-party-notices.txt file located at the root of the repository. Previously, I was also adding comments indicating the source of the ported implementation. Following your suggestion, I've now added more detailed licensing information at the top of the source files to improve clarity.

Reply:
I think there are two parts:
1. license
2. the copyright information (e.g. "// Copyright The OpenTelemetry Authors")

My understanding is that the Apache License requires you to keep the original copyright information.